app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.js

```diff
@@ -0,0 +1,119 @@
+import React, { useEffect, useCallback } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import * as actions from '../../../../redux/actions';
+import buildGenerateAGPImagesFunction from './buildGenerateAGPImagesFunction';
+
+import getOpts from './getOpts';
+import getQueries from './getQueries';
+
+export const STATUS = {  
+  // States in order of happy path AGP generation sequence
+  INITIALIZED:    'INITIALIZED',
+  STATE_CLEARED:  'STATE_CLEARED',
+  PATIENT_LOADED: 'PATIENT_LOADED',
+  DATA_PROCESSED: 'DATA_PROCESSED',
+  SVGS_GENERATED: 'SVGS_GENERATED',
+
+  // Other states
+  NO_PATIENT_DATA:   'NO_PATIENT_DATA',
+  INSUFFICIENT_DATA: 'INSUFFICIENT_DATA',
+}
+
+const inferLastCompletedStep = (patientId, data, pdf) => {
+  // If the outputted data for a step in the process exists, we infer that the step was successful.
+  // We do the lookup in reverse order to return the LATEST completed step
+
+  // Incorrect Patient --- (occurs when user switches patient partway through fetching)
+  const hasOtherPdfInState  = !!pdf.opts && pdf.opts.patient?.id !== patientId;
+  const hasOtherDataInState = !!data.metaData.patientId && data.metaData.patientId !== patientId;
+
+  if (hasOtherPdfInState || hasOtherDataInState) return STATUS.INITIALIZED;
+
+  // Insufficient Data States ---
+  const hasNoPatientData    = data.metaData?.size === 0;
+  const hasInsufficientData = !!pdf?.opts?.svgDataURLS && !pdf?.opts?.svgDataURLS.agpCGM;
+
+  if (hasNoPatientData)    return STATUS.NO_PATIENT_DATA;
+  if (hasInsufficientData) return STATUS.INSUFFICIENT_DATA;
+
+  // Happy Path States ---
+  const hasImagesInState  = !!pdf?.opts?.svgDataURLS;
+  const hasPDFDataInState = !!pdf?.data;
+  const hasPatientInState = !!data?.metaData?.patientId;
+
+  if (hasImagesInState)  return STATUS.SVGS_GENERATED;
+  if (hasPDFDataInState) return STATUS.DATA_PROCESSED;
+  if (hasPatientInState) return STATUS.PATIENT_LOADED;
+
+  return STATUS.STATE_CLEARED;
+}
```

Comment:
The decision to use this approach was really tough to make.

I also looked at using `blip.working` as an indicator of each step's success. However, in trying it out, I felt like it is intended for one-way operation of each process and there is no pre-existing mechanism to reset the values in it. Resets are needed because we need to be able to open and close the drawer multiple times for different patients. In addition, because `blip.working` does not store metadata, it was challenging to ensure that it represents the processing state for the _current_ patient instead of a different patient in the case of race conditions and such.

This approach of using relying on the data output is a bit more reliable in that sense, but it is more challenging to read and also to understand the logic of why we're making each check in the order that we are. I'm a little worried about the code being a bit brittle here too.